### PR TITLE
Waf configuration resource

### PR DESCRIPTION
**Pull request by oscarDovao**
This PR contains the implementation of the new waf version API endpoints.
Two functions provided for listing versions for a particular waf; one that takes parameter for pagination (pagination has to be done by the API's consumer) and another one which paginates through all the result set and returns the full list.

The deletion of rules when a waf configuration resources is removed will be addressed in this card https://trello.com/c/ODqSnh9Q

### CODE REVIEW

File: fastly/waf_version.go [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+// WAFVersionResponse represents a list WAF versions full response.
+type WAFVersionResponse struct {
+	Items []*WAFVersion
+	Info  infoResponse
+}
+
+// ListWAFVersionsInput used as input for listing WAF versions.
+type ListWAFVersionsInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_version.go [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+	WAFID string
```
Comment (phamann) [inline fastly/waf_version.go]: Should this be `WAFID`, or `ID` or `FirewallD` which is used in the documentation? I don't have strong opinions for which is correct, but i'd like consistency in what we choose.

File: fastly/waf_version.go [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+	}
+	return &WAFVersionResponse{
+		Items: wafVersions,
+		Info:  info,
+	}, nil
+}
+
+// ListAllWAFVersionsInput used as input for listing all WAF versions.
+type ListAllWAFVersionsInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_version.go [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+	WAFID string
+	// Include relationships. Optional, comma-separated values. Permitted values: waf_firewall_versions.
+	Include string
+}
+
+// ListAllWAFVersions returns the complete list of WAF versions for a given WAF id. It iterates through
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
// ListAllWAFVersions returns the complete list of WAF versions for a given WAF ID. It iterates through
```

File: fastly/waf_version.go [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+// all existing pages to ensure all waf versions are returned at once.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
// all existing pages to ensure all WAF versions are returned at once.
```

File: fastly/waf_version.go [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+
+		if r.Info.Links.Next == "" || len(r.Items) == 0 {
+			return result, nil
+		}
+	}
+}
+
+// GetWAFVersionInput used as input for GetWAFVersion function.
+type GetWAFVersionInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_version.go [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+	var wafVer WAFVersion
+	if err := jsonapi.UnmarshalPayload(resp.Body, &wafVer); err != nil {
+		return nil, err
+	}
+	return &wafVer, nil
+}
+
+// UpdateWAFVersionInput is used as input to the UpdateWAFVersion function.
+type UpdateWAFVersionInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_version.go [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+	WAFID string
+	// the Web Application Firewall's version id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// the Web Application Firewall's version ID.
```

File: fastly/waf_version.go [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+	var waf WAFVersion
+	if err := jsonapi.UnmarshalPayload(resp.Body, &waf); err != nil {
+		return nil, err
+	}
+	return &waf, nil
+}
+
+// LockWAFVersionInput used as input for locking a WAF version.
+type LockWAFVersionInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_version.go [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+	var waf WAFVersion
+	if err := jsonapi.UnmarshalPayload(resp.Body, &waf); err != nil {
+		return nil, err
+	}
+	return &waf, nil
+}
+
+// CloneWAFVersionInput used as input for cloning a WAF version.
+type CloneWAFVersionInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_version.go [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+	var waf WAFVersion
+	if err := jsonapi.UnmarshalPayload(resp.Body, &waf); err != nil {
+		return nil, err
+	}
+	return &waf, nil
+}
+
+// DeployWAFVersionInput used as input for deploying a WAF version.
+type DeployWAFVersionInput struct {
+	// The Web Application Firewall's id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
	// The Web Application Firewall's ID.
```

File: fastly/waf_version.go [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+		case "int":
+			if value != 0 {
+				result[key] = strconv.Itoa(value.(int))
+			}
+		}
+	}
+	return result
+}
+
+// ListWAFVersions returns the list of VAF versions for a given WAF id.
```
Comment (phamann) [inline fastly/waf_version.go]:
```suggestion
// ListWAFVersions returns the list of VAF versions for a given WAF ID.
```

File: fastly/errors.go [hunk 13]
```diff
@@ -93,10 +93,18 @@ var ErrMissingIP = errors.New("Missing requried field 'IP'")
 // requires a "EventID" key, but one was not set
 var ErrMissingEventID = errors.New("Missing required field 'EventID'")
 
-// ErrMissingWafID is an error that is returned was an input struct
-// requires a "WafID" key, but one was not set
+// ErrMissingWafID is an error that is returned when an input struct
+// requires a "WAFID" key, but one was not set
```
Comment (phamann) [inline fastly/errors.go]:
```suggestion
// requires a "WAFID" key, but one was not set.
```

File: fastly/errors.go [hunk 14]
```diff
@@ -93,10 +93,18 @@ var ErrMissingIP = errors.New("Missing requried field 'IP'")
...
 var ErrMissingWAFID = errors.New("Missing required field 'WAFID'")
 
+// ErrMissingWAFVersionNumber is an error that is returned when an input struct
+// requires a "WAFVersionNumber" key, but one was not set
```
Comment (phamann) [inline fastly/errors.go]:
```suggestion
// requires a "WAFVersionNumber" key, but one was not set.
```

File: fastly/errors.go [hunk 15]
```diff
@@ -93,10 +93,18 @@ var ErrMissingIP = errors.New("Missing requried field 'IP'")
...
+var ErrMissingWAFVersionNumber = errors.New("Missing required field 'WAFVersionNumber'")
+
+// ErrMissingWAFVersionID is an error that is returned when an input struct
+// requires a "WAFVersionID" key, but one was not set
```
Comment (phamann) [inline fastly/errors.go]:
```suggestion
// requires a "WAFVersionID" key, but one was not set.
```

File: fastly/waf_version.go [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+
+	"github.com/google/jsonapi"
+)
+
+// WAFVersionType is used for reflection because JSONAPI wants to know what it's
+// decoding into.
+var WAFVersionType = reflect.TypeOf(new(WAFVersion))
+
+// paginationPageSize used as PageSize by the ListAllWAFVersions function.
+const paginationPageSize = 100
```
[the remark this answers is not in the document]
Comment (oscarDovao) [reply fastly/waf_version.go]: The Idea was to use this for all WAF related request, so it make sense to prefix it. Renamed to WAFPaginationPageSize

File: fastly/waf_version.go (hunk 2 above)
Comment (oscarDovao) [reply fastly/waf_version.go]: We had a couple of discussions about this, and I agree whichever way we choose it should be consistent across. 
The documentation does not seem to list internal ids https://docs.fastly.com/api/ngwaf#api-section-ngwaf_firewall_versions
My preference is to use the current context (in this case WAFVersion) to keep field names shorter, which means that ID would be the WAFVersion id and any other fields that aren't directly related to WAFVersion are prefixed. However I'm happy to add full context regardless the current input context.

Comment (phamann) [reply fastly/waf_version.go]: Makes sense 👍, i'm happy to leave as is.

**Pull request merged**